### PR TITLE
Improve Markdown + Components usage

### LINKED ISSUE #2399: 🐛 BUG: Components with array-like prop fail to render in Markdown
### What version of `astro` are you using?

0.22.11

### What package manager are you using?

npm

### What operating system are you using?

Mac

### Describe the Bug

I have a component that takes an array as a prop. I then use this component in a markdown file with `setup`. When that markdown is rendered, I get something along the lines of `<ComponentName propName=[[Object object][Object object]] />` rendered a string. Interestingly, passing empty array to them component seems to work (nothing gets rendered).

I tried to create a Vue component as well and got the same result.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-fdsvay?file=src%2Fpages%2Findex.md

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- Significantly improved component + JSX expression handling in `@astrojs/markdown-remark`
  - Borrows heavily from [MDX](https://mdxjs.com/) and [`remark-mdx`](https://github.com/mdx-js/mdx/tree/main/packages/remark-mdx), but doesn't use _all_ of the features (namely inline esm imports)
  - Only caveat to callout—shorthand attributes are not currently supported. Otherwise this a MASSIVE improvement for our components-in-markdown story.
- Adds support for header slugs which contain JSX expressions
- Closes #2399, all components inside of Markdown now have correct prop serialization
- Removes need for extra newlines inside of Component before switching back to Markdown
- Removes wrapping `<p>` tags around Components

## Testing

- Adds test suite to `@astrojs/markdown-remark`
- Updates a JSX expression test in `astro` to reflect new and improved output

## Docs

TODO

### CODE REVIEW

File: packages/astro/src/vite-plugin-markdown/index.ts [hunk 1]
```diff
@@ -147,6 +152,8 @@ ${setup}`.trim();
 					site: config.site ? new URL(config.base, config.site).toString() : undefined,
 					sourcefile: id,
 					sourcemap: 'inline',
+					// TODO: baseline flag
```
Comment (tony-sull) [inline packages/astro/src/vite-plugin-markdown/index.ts:155]: just double checking, this a future PR todo correct?

File: packages/markdown/remark/src/index.ts [hunk 2]
```diff
@@ -86,8 +98,8 @@ export async function renderMarkdown(
 	try {
 		const vfile = await parser
 			.use([rehypeCollectHeaders])
-			.use(rehypeStringify, { allowDangerousHtml: true })
-			.process(content);
+			.use(rehypeStringify, { allowDangerousHtml: true, allowParseErrors: true })
```
Comment (tony-sull) [inline packages/markdown/remark/src/index.ts:101]: well that's a bit of a scary flag name 😆

Does this allow MDX syntax to get through without being thrown as bad markdown syntax?

**Review by tony-sull — COMMENTED** [review]
I'm not as familiar with some of the custom MD parsing logic updated here, but the test coverage looks good to me!

Marking this review as a comment rather than Approve in case anyone that's worked on the MD logic recently wants to chime in

File: packages/astro/src/vite-plugin-markdown/index.ts (hunk 1 above)
Comment (natemoo-re) [reply packages/astro/src/vite-plugin-markdown/index.ts:155]: Yeah, this was copy + pasted from the Astro plugin, I think a note from @matthewp

**Review by matthewp — APPROVED** [review]

**Review by bholmesdev — APPROVED** [review]
`REMARK`ABLE WORK!

**Comment by aFuzzyBear** [comment]
I have one question. Would this allow for components in markdown files that are placed elsewhere outside of the `src/pages/` to work as well. So for those projects that have a top-level docs folder in there monorepo setup, with components being relative to the markdown file, would they be able to now benefit from having components in markdown or are they still plain markdown files?

Otherwise this looks amazing, great work dude 💚💚💚
(Kinda have this thing where markdown shouldnt be in the UI )

**Comment by natemoo-re** [comment]
> I have one question. Would this allow for components in markdown files that are placed elsewhere outside of the `src/pages/` to work as well?

These changes don't touch our routing logic, just the markdown processing itself. Please open an issue detailing your use case so we can track this separately!

**Pull request merged**